### ISSUE TITLE
OOM when running hack/verify-typecheck.sh with go 1.17

### ISSUE THREAD

**Issue by danmillwood** [kubernetes/kubernetes#105681]
### What happened?

go 1.17 appears to have significantly increased the amount of memory required on a kubernetes development environment to run `make verify` and in particular, `hack/verify-typecheck.sh` `verify-typecheck-dockerless.sh` and `verify-typecheck-providerless.sh`

With go 1.16.9, these can be run in a Linux VM with 8Gb memory.  In go 1.17, `hack/verify-typecheck.sh` still fails with 16Gb memory.

### What did you expect to happen?

The Kubernetes development guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md recommends 8GB of RAM to build Kubernetes.

The README file https://github.com/kubernetes/kubernetes/blob/master/test/typecheck/README states:

```
Typecheck does cross-platform typechecking of source code for all Go build
platforms.

The primary benefit is speed: a full Kubernetes cross-build takes 20 minutes
and >40GB of RAM, while this takes under 2 minutes and <8GB of RAM. 
```

I expected to be able to run these on my Linux VM with 8 GB RAM.

### How can we reproduce it (as minimally and precisely as possible)?

On a Linux VM with 8GB RAM
Install go 1.17.2
run `hack/verify-typecheck.sh`

```
root@kubernetes1804:~/go/src/k8s.io/kubernetes# hack/verify-typecheck.sh
Generating prerelease lifecycle code for 6 targets
Generating deepcopy code for 36 targets
Generating defaulter code for 13 targets
Generating conversion code for 19 targets
Generating openapi code for KUBE
Generating openapi code for AGGREGATOR
Generating openapi code for APIEXTENSIONS
Generating openapi code for CODEGEN
Generating openapi code for SAMPLEAPISERVER
type-checking windows/arm64
type-checking linux/arm
fatal error: runtime: out of memory

runtime stack:
runtime.throw({0x6205fa, 0x800000})
	/usr/local/go/src/runtime/panic.go:1198 +0x71
runtime.sysMap(0xc2c5800000, 0x428c40, 0xc1688bbe90)
	/usr/local/go/src/runtime/mem_linux.go:169 +0x96
runtime.(*mheap).grow(0x7bb680, 0x242)
	/usr/local/go/src/runtime/mheap.go:1393 +0x225
runtime.(*mheap).allocSpan(0x7bb680, 0x242, 0x0, 0x0)
	/usr/local/go/src/runtime/mheap.go:1179 +0x165
runtime.(*mheap).alloc.func1()
	/usr/local/go/src/runtime/mheap.go:913 +0x69
runtime.systemstack()
	/usr/local/go/src/runtime/asm_amd64.s:383 +0x49
```

```
root@kubernetes1804:~/go/src/k8s.io/kubernetes# cat /proc/meminfo 
MemTotal:        8167568 kB
MemFree:         7849560 kB
MemAvailable:    7812628 kB
```

### Anything else we need to know?

I've run exactly the same code level against `go 1.16.9` and `go 1.17.2`. The first works, the second doesn't.

A similar problem was spotted here https://github.com/kubernetes/test-infra/issues/23374 where they reduced the number of checks done in parallel, but this done not solve the issue for developers.

Increasing swap memory helps, but it takes several hours to run.

I wondered if the first action might be to raise an issue with the golang team and see what they think?  Id be happy to do that.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.0-alpha.3.316+d09ce7be1c3b98-dirty", GitCommit:"d09ce7be1c3b98afabaab1e01494546a2f4a95c5", GitTreeState:"dirty", BuildDate:"2021-10-14T15:25:02Z", GoVersion:"go1.17.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.0-alpha.3.316+d09ce7be1c3b98-dirty", GitCommit:"d09ce7be1c3b98afabaab1e01494546a2f4a95c5", GitTreeState:"dirty", BuildDate:"2021-10-14T15:25:02Z", GoVersion:"go1.17.2", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
N/A
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

$ uname -a
Linux kubernetes1804 4.15.0-159-generic #167-Ubuntu SMP Tue Sep 21 08:55:05 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```
</details>

### Install tools

<details>
git clone
</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by BenTheElder** [comment]
Appears to be a valid upstream issue in go 👀  (and probably fixed with go1.18?) unfortunately I don't think existing Kubernetes branches will upgrade a minor go version, only patches, but future work can switch ...

**Comment by BenTheElder** [comment]
/remove-lifecyle rotten
we're on go1.18 now, perhaps this will be resolved?

**Comment by aojea** [comment]
> /remove-lifecyle rotten we're on go1.18 now, perhaps this will be resolved?

I think that go1.18 consumes more memory https://github.com/kubernetes/kubernetes/issues/108357#issuecomment-1053970948

**Comment by BenTheElder** [comment]
> I think that go1.18 consumes more memory https://github.com/kubernetes/kubernetes/issues/108357#issuecomment-1053970948

That's talking about go at runtime, but this issue involved a large increase in the compiler code specifically, which should be resolved by removing the problematic code in 1.18 https://github.com/golang/go/issues/49035#issuecomment-947887007

**Comment by dims** [comment]
closing, please see https://github.com/golang/go/issues/49035#issuecomment-1209607751

**Comment by BenTheElder** [comment]
I think we’re still running at reduced parallelism unfortunately. Might be worth checking if we can improve that yet
